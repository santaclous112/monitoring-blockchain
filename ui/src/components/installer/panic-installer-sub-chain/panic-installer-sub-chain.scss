@import '../../scss/mixins/centralized-container.scss';
@import '../../scss/mixins/heading.scss';

.panic-installer-sub-chain {

  &__container {
    padding: 5px 5% 5px 5%;

    svc-surface {
      margin: 10px;
    }
  }

  &__heading {
    @include heading-container();
  }

  &__title {
    @include heading-title();
  }

  &__step {
    @include heading-step();
  }

  &__message-container {
    display: flex;
    align-items: center;
  }

  &__message {
    font-size: 17px;
    color: var(--ion-color-primary);
  }

  &__list {
    margin-top: -10px;
    font-size: 17px;
    color: var(--ion-color-primary);
  }

  &__list-item {
    margin-bottom: 5px;
  }

  form {

    #submit {
      display: none;
    }
  }

  svc-input {
    margin: 10px 2px 10px 2px;
  }

  svc-select {

    ion-select {
      margin: 1px;
    }
  }
}

// setting info modal dimension
.panic-installer-sub-chain {
  &__info-modal {
    --height: 690px !important;
    --width: 800px !important;
  }
}

.alert-radio-label {
  text-transform: capitalize;
}

#unsaved-changes-text {
  color: var(--ion-color-tertiary);
  font-weight: 600;
  margin-top: auto;
}