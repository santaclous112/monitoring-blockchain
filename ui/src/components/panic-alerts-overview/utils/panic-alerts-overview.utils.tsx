import {h} from "@stencil/core";
import {Alert} from "../../../interfaces/alerts";
import {SubChain, Source} from "../../../interfaces/chains";
import {
    DataTableHeaderType,
    DataTableRecordType
} from "@simply-vc/uikit/dist/types/types/datatable";
import {SelectOptionType} from "@simply-vc/uikit/dist/types/types/select";
import {HelperAPI} from "../../../utils/helpers";
import {SeverityAPI} from "../../../utils/severity";
import {ALERTS_DATA_TABLE_HEADER} from "../../../utils/constants";

export const AlertsOverviewAPI = {
    getChainFilterOptionsFromChains: getChainFilterOptionsFromChains,
    getSourcesFilterOptions: getSourcesFilterOptions,
    getDataTableJSX: getDataTableJSX,
    validateDateTimeOrder: validateDateTimeOrder,
    formatEventData: formatEventData
}

/**
 * Formats sub-chains to {@link SelectOptionType}.
 * @param subChains array of sub-chains to be formatted.
 * @returns populated {@link SelectOptionType} object.
 */
function getChainFilterOptionsFromChains(subChains: SubChain[]): SelectOptionType {
    return subChains.map(chain => ({label: chain.name, value: chain.name}));
}

/**
 * Formats sources to {@link SelectOptionType}.
 * @param sources array of sources to be formatted.
 * @returns populated {@link SelectOptionType} object.
 */
function getSourcesFilterOptions(sources: Source[]): SelectOptionType {
    return sources.map(source => ({label: source.name, value: source.id}));
}

/**
 * Gets the JSX for the data table.
 * @param alerts array of alerts to be displayed.
 * @returns populated data table as {@link JSX.Element}.
 */
function getDataTableJSX(alerts: Alert[]): JSX.Element {
    let hasAlerts = alerts.length > 0;
    const cols: DataTableHeaderType[] = ALERTS_DATA_TABLE_HEADER;
    const rows: DataTableRecordType = hasAlerts ? getDataTableRecordTypeFromAlerts(alerts) : [];
    hasAlerts = rows.length > 0;

    return <svc-data-table
        key={hasAlerts ? 'data-table-no-alerts' : 'data-table-alerts'}
        cols={cols}
        rows={rows}
        last-clicked-column-index={1}
        ordering={"descending"}
        no-records-message={"There are no alerts to display at this time"}
    />
}

/**
 * Gets {@link DataTableRecordType} from an array of alerts.
 * @param alerts array of alerts.
 * @returns populated {@link DataTableRecordType} object.
 */
function getDataTableRecordTypeFromAlerts(alerts: Alert[]): DataTableRecordType {
    return alerts.map(alert => [
        {
            label: SeverityAPI.getSeverityIcon(alert.severity),
            value: alert.severity
        },
        {
            label: new Date(alert.timestamp * 1000).toLocaleString(),
            value: new Date(alert.timestamp * 1000)
        },
        {label: alert.message, value: alert.message}]);
}

/**
 * Validates the 'from' and 'to' datetime values where the 'from' datetime value
 * should be smaller than or equal to the 'to' datetime value.
 * @param from from datetime value in string format.
 * @param to to datetime value in string format.
 * @returns true if valid, false if invalid.
 */
function validateDateTimeOrder(from: string, to: string): boolean {
    // Return true if any of the date/time values is not set.
    if (from === '' || to === '') {
        return true;
    }

    return HelperAPI.dateTimeStringToTimestamp(from) <= HelperAPI.dateTimeStringToTimestamp(to);
}

/**
 * Formats the event data generated by the svc-filter/svc-select components.
 * @param eventData event data in string format.
 * @returns formatted event data as an array of strings.
 */
function formatEventData(eventData: string): string[] {
    // Split selected options from event data string using commas.
    const selectedOptions: string[] = eventData.split(',');

    // Remove empty string element from array if no options are selected.
    if (selectedOptions.length > 0 && selectedOptions[0] === '') {
        selectedOptions.pop();
    }

    return selectedOptions;
}
