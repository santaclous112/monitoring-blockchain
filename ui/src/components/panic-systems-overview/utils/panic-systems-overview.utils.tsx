import {h} from "@stencil/core";
import {MetricAlert, Metrics} from "../../../interfaces/metrics";
import {
    DataTableHeaderType, DataTableRowsType, DataTableRowCellType
} from "@simply-vc/uikit/dist/types/types/datatable";
import {
    SystemMetricKeys,
    Severity,
    SYSTEM_METRICS_DATA_TABLE_HEADER
} from "../../../utils/constants";

export const SystemsOverviewAPI = {
    getDataTableJSX: getDataTableJSX,
    checkIfAnyNotAvailable: checkIfAnyNotAvailable,
}

/**
 * Gets the JSX for the data table.
 * @param metrics array of metrics.
 * @returns populated data table as {@link JSX.Element}.
 */
function getDataTableJSX(metrics: Metrics[]): JSX.Element {
    const cols: DataTableHeaderType[] = SYSTEM_METRICS_DATA_TABLE_HEADER;
    const rows: DataTableRowsType = getDataTableRowsTypeFromMetrics(metrics);
    let hasMetrics: boolean = metrics.length > 0;

    return <svc-data-table
        key={hasMetrics ? 'data-table-no-alerts' : 'data-table-alerts'}
        cols={cols}
        rows={rows}
        last-clicked-column-index={0}
        ordering={'ascending'}
        no-records-message="There are no metrics to display at this time"
    />
}

/**
 * Gets {@link DataTableRowsType} from an array of metrics.
 * @param metrics array of metrics.
 * @returns populated {@link DataTableRowsType} object.
 */
function getDataTableRowsTypeFromMetrics(metrics: Metrics[]): DataTableRowsType {
    return metrics.map(metric => ({
        cells: [
            {label: metric.name, value: metric.id},
            metricToCellType(metric.metrics.CPU, SystemMetricKeys.SYSTEM_CPU_USAGE, metric.metricAlerts),
            metricToCellType(metric.metrics.RAM, SystemMetricKeys.SYSTEM_RAM_USAGE, metric.metricAlerts),
            metricToCellType(metric.metrics.Storage, SystemMetricKeys.SYSTEM_STORAGE_USAGE, metric.metricAlerts),
            metricToCellType(metric.metrics.ProcessMemory),
            metricToCellType(metric.metrics.ProcessVirtualMemory),
            metricToCellType(metric.metrics.OpenFileDescriptors, SystemMetricKeys.OPEN_FILE_DESCRIPTORS, metric.metricAlerts)
        ],
        id: metric.id
    }));
}

/**
 * Formats metrics to {@link DataTableRowCellType}. If the metric passed is a
 * number, the cell's label is succeeded by a percentage. If the metric key and
 * metric alerts are passed, the css classes field is populated accordingly
 * using {@link checkMetricAlerts}.
 * @param metricValue the value of the metric, can be either a number or a string.
 * @param metricKey the metric key of the passed metric, optional.
 * @param metricAlerts array of metric alerts, optional.
 * @returns populated {@link DataTableRowCellType} object.
 */
function metricToCellType(metricValue: number | string, metricKey?: SystemMetricKeys, metricAlerts?: MetricAlert[]): DataTableRowCellType {
    if (metricValue === undefined || metricValue === null || Number.isNaN(metricValue)) {
        return {label: 'N/A', value: null, cssClasses: ['not-available']};
    }

    const cellType: DataTableRowCellType = {
        label: typeof metricValue === 'number' ? `${metricValue}%` : metricValue,
        value: metricValue
    };

    if (metricKey && metricAlerts) {
        cellType.cssClasses = [checkMetricAlerts(metricKey, metricAlerts)];
    }

    return cellType;
}

/**
 * Returns the status of a given metric based on an array of metric alerts.
 * @param metric metric key being checked.
 * @param metricAlerts array of metric alerts.
 * @returns string with severity of metric or 'INFO' severity if no metric alert
 * is found.
 */
function checkMetricAlerts(metric: SystemMetricKeys, metricAlerts: MetricAlert[]): string {
    const metricAlert: MetricAlert = metricAlerts.find(
      alert => SystemMetricKeys[alert.metric] === metric);

    if (metricAlert !== undefined) {
        return Severity[metricAlert.severity].toLowerCase();
    }

    return Severity.INFO.toLowerCase();
}

/**
 * Checks whether any metric is null (not available).
 * @param metrics array of metrics.
 * @returns true if any metric is null, otherwise returns false.
 */
function checkIfAnyNotAvailable(metrics: Metrics[]): boolean {
    return metrics.some((systemMetric) => {
        return Object.values(systemMetric.metrics).some(
          metric => metric === undefined || metric === null);
    });
}
