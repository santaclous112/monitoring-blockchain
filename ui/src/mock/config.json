[
  {
    "id": "62675e2d8891cd77b87f5b16",
    "status": true,
    "created": "2022-05-02T22:02:46.000Z",
    "modified": null,
    "baseChain": {
      "id": "6265cefcfdb17d641746dced",
      "status": true,
      "name": "Cosmos",
      "value": "cosmos",
      "description": "Cosmos description here...",
      "sources": [
        {
          "id": "6271178cf740ef3847a7d27e",
          "status": true,
          "created": "2022-04-01T00:00:00.000Z",
          "modified": null,
          "name": "cosmos_node_1",
          "value": null,
          "description": null
        }
      ]
    },
    "subChain": {
      "id": "626aeb43980cf43ee7bbf683",
      "status": true,
      "created": "2022-05-02T22:02:46.000Z",
      "modified": null,
      "name": "Akash"
    }
  },
  {
    "id": "62675e2d8891cd77b87f5b166",
    "status": true,
    "created": "2022-05-02T22:02:46.000Z",
    "modified": null,
    "baseChain": {
      "id": "6265cefcfdb17d641746dcedd",
      "status": true,
      "name": "Cosmos",
      "value": "cosmos",
      "description": "Cosmos description here...",
      "sources": [
        {
          "id": "6271178cf740ef3847a7d27e",
          "status": true,
          "created": "2022-04-01T00:00:00.000Z",
          "modified": null,
          "name": "cosmos_node_2",
          "value": null,
          "description": null
        }
      ]
    },
    "subChain": {
      "id": "626aeb43980cf43ee7bbf6833",
      "status": true,
      "created": "2022-05-02T22:02:46.000Z",
      "modified": null,
      "name": "Akash 2"
    }
  }
]