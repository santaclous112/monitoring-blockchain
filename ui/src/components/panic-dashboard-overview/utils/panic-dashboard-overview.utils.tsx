import { h } from '@stencil/core';
import { Alert, Severity } from '../../../interfaces/alerts';
import { BaseChain } from '../../../interfaces/chains';
import { DataTableRecordType } from '../../../lib/types/types/datatable';
import { OrderingType } from '../../../lib/types/types/ordering';
import { SelectOptionType } from '../../../lib/types/types/select';
import { AlertsAPI } from '../../../utils/alerts';

export const DashboardOverviewAPI = {
    getChainFilterOptionsFromBaseChain: getChainFilterOptionsFromBaseChain,
    getPieChartJSX: getPieChartJSX,
    getDataTableJSX: getDataTableJSX
}

/**
 * Formats base chain to SelectOptionType type.
 * @param baseChain base chain to be converted.
 * @returns populated list of required object type.
 */
function getChainFilterOptionsFromBaseChain(baseChain: BaseChain): SelectOptionType {
    return baseChain.chains.map(chain => ({ label: chain.name, value: chain.name }))
}

/**
 * Gets the JSX for the pie chart.
 * @param chainName the respective chain's name.
 * @param criticalAlerts number of critical alerts.
 * @param warningAlerts number of warning alerts.
 * @param errorAlerts number of error alerts.
 * @returns populated pie chart JSX.
 */
function getPieChartJSX(baseChain: BaseChain): JSX.Element {
    let criticalAlerts: number = 0;
    let warningAlerts: number = 0;
    let errorAlerts: number = 0;

    for (const chain of baseChain.chains) {
        if (chain.active) {
            criticalAlerts += chain.alerts.filter(alert => Severity[alert.severity] === Severity.CRITICAL).length;
            warningAlerts += chain.alerts.filter(alert => Severity[alert.severity] === Severity.WARNING).length;
            errorAlerts += chain.alerts.filter(alert => Severity[alert.severity] === Severity.ERROR).length;
        }
    }

    const hasAlerts: boolean = criticalAlerts + warningAlerts + errorAlerts > 0;
    // PieChart config with alerts
    const cols = [{ title: 'Alert', type: 'string' }, { title: 'Amount', type: 'number' }];
    const rows = [['Critical', criticalAlerts], ['Warning', warningAlerts], ['Error', errorAlerts]];
    const alertsColors: string[] = ['#a39293', '#f4dd77', '#f7797b'];
    // PieChart config without alerts
    const noAlertsRows = [['', 1]];
    const noAlertsColors: string[] = ['#b0ea8f'];

    return <svc-pie-chart
        key={hasAlerts ? `${baseChain.name}-pie-chart-no-alerts` : `${baseChain.name}-pie-chart-alerts`}
        colors={hasAlerts ? alertsColors : noAlertsColors}
        cols={cols}
        rows={hasAlerts ? rows : noAlertsRows}
        pie-slice-text={hasAlerts ? "percentage" : "none"}
        tooltip-trigger={hasAlerts ? "focus" : "none"}
    />
}

/**
 * Gets the JSX for the data table.
 * @param alerts list of alerts to be displayed.
 * @returns populated data table JSX.
 */
function getDataTableJSX(baseChain: BaseChain): JSX.Element {
    let alerts: Alert[] = [];

    for (const chain of baseChain.chains) {
        if (chain.active) {
            alerts.push.apply(alerts, chain.alerts);
        }
    }

    const hasAlerts = alerts.length > 0;
    const cols: string[] = ['Severity', 'Time Stamp', 'Message'];
    const rows: DataTableRecordType = hasAlerts ? getDataTableRecordTypeFromAlerts(alerts, baseChain.activeSeverities) : [];

    return <svc-data-table
        id={baseChain.name}
        key={hasAlerts ? `${baseChain.name}-data-table-no-alerts` : `${baseChain.name}-data-table-alerts`}
        cols={cols}
        rows={rows}
        ordering={baseChain.ordering as OrderingType}
        last-clicked-column-index={baseChain.lastClickedColumnIndex}
        no-records-message="There are no alerts to display at this time"
    />
}

/**
 * Filters alerts which do not have an active severity and formats alerts to DataTableRecordType type.
 * @param alerts list of alerts.
 * @param activeSeverities list of active severities.
 * @returns populated list of lists of required object type.
 */
function getDataTableRecordTypeFromAlerts(alerts: Alert[], activeSeverities: Severity[]): DataTableRecordType {
    // Filter alerts.
    const filteredAlerts = alerts.filter(function (alert) {
        return activeSeverities.includes(alert.severity);
    });

    // Format filtered alerts into DataTableRecordType type.
    return filteredAlerts.map(alert => [
        { label: AlertsAPI.getSeverityIcon(alert.severity), value: alert.severity },
        { label: new Date(alert.timestamp * 1000).toLocaleString(), value: new Date(alert.timestamp * 1000) },
        { label: alert.message, value: alert.message }]);
}
