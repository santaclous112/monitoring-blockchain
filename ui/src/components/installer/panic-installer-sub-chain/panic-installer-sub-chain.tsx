import {createAlert, createInfoAlert, createModal, dismissModal, parseForm} from '@simply-vc/uikit';
import {SelectOptionType} from '@simply-vc/uikit/dist/types/types/select';
import {Component, h, Host, Listen, Prop, State} from '@stencil/core';
import {HelperAPI} from '../../../utils/helpers';
import {
  BULLET_ONE,
  BULLET_TWO,
  MAIN_TEXT,
  MORE_INFO_MESSAGES
} from '../content/sub-chain';
import {Config} from '../../../../../entities/ts/Config';
import {ConfigService} from '../../../services/config/config.service';
import {BaseChain} from '../../../../../entities/ts/BaseChain';
import {Router} from "stencil-router-v2";
import { CONFIG_ID_LOCAL_STORAGE_KEY } from '../../../utils/constants';

@Component({
  tag: 'panic-installer-sub-chain',
  styleUrl: 'panic-installer-sub-chain.scss'
})
export class PanicInstallerSubChain {

  /**
   * Stencil Router object for page navigation.
   */
  @Prop() router: Router;

  /**
   * List of blockchain frameworks supported by PANIC.
   */
  @Prop() baseChains: BaseChain[] = [];

  /**
   * Indicates whether the installer process is editing a previously created
   * config.
   */
  @Prop() configId: string;

  /**
   * Stores the *current* base chain being configured. Useful to pre-fill the
   * form when the node operator navigates back to sub-chain page.
   */
  @State() currBaseChain: BaseChain;

  /**
   * The config to be edited.
   */
  @State() config: Config;

  /**
   * The current text within the sub chain name field.
   */
  @State() subChainName: string;

  /**
   * Whether a general chain exists among the user's configs.
   */
  generalChainExists: boolean;

  /**
   * Converts the list of base chains to the {@link SelectOptionType} format.
   *
   * @returns list of base chains in the {@link SelectOptionType} format.
   */
  baseChainToSelectOptionType(): SelectOptionType {

    let baseChainsToShow = [...this.baseChains];

    if(this.config){
      if(this.generalChainExists && this.config.baseChain.value !== 'general'){
        baseChainsToShow = baseChainsToShow.filter(chain => chain.value !== "general");
      }
    }

    return baseChainsToShow.map((baseChain: BaseChain) => {
      return {
        label: baseChain.name,
        value: baseChain.id
      }
    });
  }

  async componentWillLoad() {
    // might be necessary
    // when the node operator opens a modal and then clicks in the browser's back button
    dismissModal();

    this.generalChainExists = await this.hasGeneralChain();

    // check both URL and local storage for the configID
    this.configId = this.configId ? this.configId : localStorage.getItem(CONFIG_ID_LOCAL_STORAGE_KEY);

    if (this.configId) {
      await this.refreshConfig();
      if(this.config){
        this.currBaseChain = this.config.baseChain;
        this.subChainName = this.config.subChain.name
      } else {
        localStorage.removeItem('configId');
        createInfoAlert({
          header: "Attention",
          message: "Invalid configuration. Redirecting to PANIC home page.",
          eventName: "errorReturnToHome"
        });
      }
    }
  }

  @Listen("errorReturnToHome", {target: 'window'})
  async errorReturnToHomeHandler() {
    HelperAPI.changePage(this.router, '/');
  }

  /**
   * Fetches the config data from DB.
   */
  async refreshConfig() {
    this.config = await ConfigService.getInstance().getByID(this.configId);
  }

  /**
   * If the node operator changes the base chain we need to display a dialog
   * warning about the config loss in case that the action is confirmed.
   *
   * @param selectedBaseChainId new base chain selected
   */
  handleOnChange(selectedBaseChainId: string) {
    const baseChain: BaseChain = this.baseChains.find(baseChain => baseChain.id === selectedBaseChainId);
    if (this.configId && this.currBaseChain?.id !== selectedBaseChainId) {
      createAlert({
        header: "Attention",
        message: "If you change the base chain all current configuration " +
            "will be lost. Would you like to proceed?",
        eventName: "baseChainChange",
        eventData: baseChain
      });
    } else {
      this.currBaseChain = baseChain;
    }
  }

  /**
   * This function handles the event fired from the base chain change dialog.
   *
   * @param e {@link CustomEvent} with a boolean as argument indicating
   * whether the action has been confirmed.
   */
  @Listen("baseChainChange", {
    target: 'window'
  })
  async onBaseChainChange(e: CustomEvent) {
    const changeConfirmed = e.detail.confirmed;
    const selectedBaseChain = e.detail.data;

    if (changeConfirmed) {
      // If the base chain was changed we need to delete the config from the
      // database and adjust the state. Note that we can only reach this point
      // if a configuration has already been saved, as otherwise the modal would
      // not have been triggered

      if(this.config.baseChain.value === 'general'){
        this.generalChainExists = false;
      }

      const success: boolean = await ConfigService.getInstance().delete(
          this.configId);

      // We want to execute this logic only if the delete call is successful.
      // Otherwise, we will raise a feedback message
      const toastMsg = "Could not delete configuration from database. " +
          "Basechain change failed."
      const callback = function () {
        this.currBaseChain = selectedBaseChain;
        this.config = undefined;
        this.configId = undefined;
      }.bind(this);
      await HelperAPI.executeWithFailureFeedback(
          success, callback, toastMsg, 3000, "danger")

      if (success) {
        localStorage.removeItem('configId');
        // We do not want to retrieve the previous value of svc-select, so in
        // this case return
        return;
      }
    }

    // If the user decides not to proceed with the change or the delete call
    // fails then it is necessary to keep the correct value for svc-select
    // (displayed in the component UI)
    const baseChainSelect = document.getElementById("base_chain").children[0] as HTMLInputElement;
    baseChainSelect.value  = this.currBaseChain?.id;
  }

  /**
   * Used to navigate back to the Welcome page in the installer.
   */
  @Listen("subChainNameChange", {target: "window"})
  subChainNameChangeHandler(e: CustomEvent) {
    this.subChainName = e.detail.name;
  }

  /**
   * Logic to update the sub-chain name in the DB.
   */
  async subChainNameChange() {
    const reqBody = {
      "id": this.configId,
      "subChain": {
        "name": this.subChainName
      }
    } as Config

    const response: Response = await ConfigService.getInstance().save(reqBody);

    if (HelperAPI.isDuplicateName(response)) {
      HelperAPI.raiseToast(
        `A sub-chain named '${reqBody.subChain.name}' already exists!`,
        2000,
        "warning");
    } else {
      // We want to execute this logic only if the save call is successful.
      // Otherwise, we will raise a feedback message
      const toastMsg = "Error saving configuration.";
      const callback = function () {
        HelperAPI.raiseToast(
          `Sub-chain name saved!`,
          2000,
          "success");
      }.bind(this);

      await this.refreshConfig();

      await HelperAPI.executeWithFailureFeedback(response.ok, callback, toastMsg, 3000, "danger");
    }
  }


  /**
   * Used to navigate back to the Welcome page in the installer.
   */
  @Listen("previousStep", {target: "window"})
  previousStepHandler() {
    HelperAPI.changePage(this.router, "/installer/welcome");
  }

  /**
   * Triggers a click in the `submit` button which is part of the sub-chain
   * form
   */
  @Listen("nextStep", {target: "window"})
  async nextStepHandler() {
      const button = document.getElementById("submit").children[0] as HTMLButtonElement;
      button.click();
  }

  /**
   * Given a parsed form this function will create a new config object and save
   * it in the database
   * @param parsedForm the form parsed using HelperAPI.parseForm
   */
  async createAndSendNewConfig(parsedForm: any) {
    const reqBody = {
      "id": this.configId,
      "baseChain": {
        "id": parsedForm.base_chain
      },
      "subChain": {
        "name": parsedForm.name
      }
    } as Config

    const response: Response = await ConfigService.getInstance().save(reqBody);

    if(HelperAPI.isDuplicateName(response)) {
      HelperAPI.raiseToast(
          `A sub-chain named '${reqBody.subChain.name}' already exists!`,
          2000,
          "warning");
    } else {
      // We want to execute this logic only if the save call is successful.
      // Otherwise, we will raise a feedback message
      const toastMsg = "Error saving configuration.";
      const callback = async function () {
        const configID: string = (await response.json()).result;
        HelperAPI.changePage(this.router, `${HelperAPI.getUrlPrefix()}/channels/${configID}`);
        localStorage.setItem(CONFIG_ID_LOCAL_STORAGE_KEY, configID);
      }.bind(this);

      await HelperAPI.executeWithFailureFeedback(response.ok, callback, toastMsg, 3000, "danger");
    }
  }

  /**
   * This method is fired when the node operator clicks in the "next" button.
   *
   * It will parse the form's data and make an API call to save the config.
   *
   * @param e the form event
   */
  async handleSubmit(e: Event) {
    e.preventDefault();
    const form = e.target as HTMLFormElement;
    const parsedForm: any = parseForm(form);
    if (!parsedForm.base_chain) {
      HelperAPI.raiseToast("Please, choose a base chain to proceed.", 2500, "warning");
      return;
    }

    await this.createAndSendNewConfig(parsedForm);
  }

  /**
   * Check if there is at least one General chain configured.
   *
   * @returns `true` if there is one, otherwise, `false`.
   */
  async hasGeneralChain(): Promise<boolean> {
    const configs: Config[] = await ConfigService.getInstance().getAll();
    const hasGeneralChain: boolean = configs.some((config: Config) => {
      return config.baseChain.name.toLowerCase() === "general";
    });

    return hasGeneralChain;
  }

  /**
   * Checks if the config being edited belongs to a General base chain.
   * 
   * @returns `true` if it belongs, otherwise `false`.
   */
  isGeneral(): boolean {
    const generalConfig: boolean = this.config?.baseChain.value === 'general';
    const generalCurrBaseChain: boolean = this.currBaseChain?.value === 'general';
    return generalConfig || generalCurrBaseChain;
  }

  /**
   * Returns the appropriate `placeholder` for the base chain input.
   * 
   * @returns a `string` representing the `placeholder`.
   */
  getBaseChainPlaceholder(): string {
    return this.config ? this.config.baseChain.name : "Select a base chain...";
  }

  render() {
    // show svc-input disabled when editing a chain only (Settings/Chain "edit mode")
    const baseChainInput = this.config && HelperAPI.isFromSettingsEdit()
    ?  <svc-input value={this.config.baseChain.name} disabled={HelperAPI.isFromSettingsEdit()} />
    : <svc-select
        header={"Base-chains"}
        options={this.baseChainToSelectOptionType()}
        name={"base_chain"}
        id={"base_chain"}
        value={this.currBaseChain?.id}
        withBorder={true}
        placeholder={this.getBaseChainPlaceholder()}
        // @ts-ignore
        onIonChange={(e: CustomEvent) => {
          this.handleOnChange(e.detail.value);
        }}
      />
    
    const subChainName: string = this.isGeneral() ? 'GENERAL' : this.config?.subChain.name;

    return (
      <Host class={"panic-installer-sub-chain"}>
        <panic-header showMenu={false} />

        <svc-progress-bar value={0.1} color={"tertiary"} />

        <svc-content-container class={"panic-installer-sub-chain__container"}>
          <svc-surface>
            <div class={"panic-installer-sub-chain__heading"}>
              <svc-icon color={"primary"} name={"link"} size={"100px"} />
              <svc-label
                  position={"center"}
                  class={"panic-installer-sub-chain__title"}
              >
                Blockchain Setup
              </svc-label>
              <svc-label
                  position={"center"}
                  class={"panic-installer-sub-chain__step"}
              >
                step 1/5
              </svc-label>
            </div>

            <div class={"panic-installer-sub-chain__message-container"}>
              <p class={"panic-installer-sub-chain__message"}>
                { MAIN_TEXT }
              </p>
              <svc-buttons-container>
                  <svc-button
                      color={"secondary"}
                      iconName={"information-circle"}
                      iconPosition={"icon-only"}
                      onClick={() => { createModal("panic-installer-more-info-modal", {messages: MORE_INFO_MESSAGES}, {cssClass: "panic-installer-sub-chain__info-modal"});
                      }}
                  />
              </svc-buttons-container>
            </div>

            <p class={"panic-installer-sub-chain__message"}>
              <ul class={"panic-installer-sub-chain__list"}>
                <li class={"panic-installer-sub-chain__list-item"}>
                  { BULLET_ONE }
                </li>
                <li>
                  { BULLET_TWO }
                </li>
              </ul>
            </p>

            <form onSubmit={
              (e) => {this.handleSubmit(e)}}
            >
              { baseChainInput }

              { this.configId && <input type={"hidden"} value={this.config.baseChain.id} name={"base_chain"} /> }

              { this.currBaseChain && this.currBaseChain.value === 'general' && <input type={"hidden"} value={'GENERAL'} name={"name"} /> }

              {
                !HelperAPI.isFromSettingsEdit()
                ?
                  <svc-input
                    id={"subChainName"}
                    value={subChainName}
                    disabled={this.isGeneral()}
                    type={"text"}
                    name={"name"}
                    lines={"full"}
                    label={"Name"}
                    placeholder={"Ex: Akash, Polkadot"}
                    required
                  />
                :
                  <div>
                    <svc-event-emitter
                      eventName={"subChainNameChange"}
                      debounce={300}
                    >
                      <svc-input
                        id={"subChainName"}
                        value={subChainName}
                        disabled={this.isGeneral()}
                        type={"text"}
                        name={"name"}
                        lines={"full"}
                        label={"Name"}
                        placeholder={"Ex: Akash, Polkadot"}
                        required
                      />
                    </svc-event-emitter>
                    {
                      this.subChainName !== this.config.subChain.name &&
                      <p id={"unsaved-changes-text"}>
                        There are unsaved changes!
                      </p>
                    }
                    <svc-button
                      onClick={()=>this.subChainNameChange()}
                    >
                      Save
                    </svc-button>
                  </div>
              }

              <svc-button type={"submit"} id={"submit"} />
            </form>
          </svc-surface>

          <panic-installer-nav config={this.config} hidePreviousBtn={HelperAPI.isFromSettings()} />
        </svc-content-container>

        <panic-footer />
      </Host>
    );
  }
}