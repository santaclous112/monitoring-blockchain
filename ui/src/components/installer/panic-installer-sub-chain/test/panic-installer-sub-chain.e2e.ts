import { newSpecPage } from '@stencil/core/testing';
import { PanicInstallerSubChain } from '../panic-installer-sub-chain';
import {ConfigService} from "../../../../services/config/config.service";
import {mockGetAllConfigs} from "../../../../utils/mock";

beforeAll(() => {
  jest.spyOn(ConfigService.prototype, 'getAll').mockImplementation(
    (_: string) => mockGetAllConfigs()
  );
});

afterAll(() => {
  jest.restoreAllMocks();
});

describe('panic-installer-sub-chain', () => {
  it('renders', async () => {
      const page = await newSpecPage({
          components: [PanicInstallerSubChain],
          html: '<panic-installer-sub-chain></panic-installer-sub-chain>',
      });

      const panicInstallerSubChain = page.body.querySelector('panic-installer-sub-chain');
      expect(panicInstallerSubChain).toMatchSnapshot();
  });
});