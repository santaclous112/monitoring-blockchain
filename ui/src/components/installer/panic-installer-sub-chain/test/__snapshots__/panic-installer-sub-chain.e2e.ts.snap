// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`panic-installer-sub-chain renders 1`] = `
<panic-installer-sub-chain class="panic-installer-sub-chain">
  <panic-header></panic-header>
  <svc-progress-bar color="tertiary" value="0.1"></svc-progress-bar>
  <svc-content-container class="panic-installer-sub-chain__container">
    <svc-surface>
      <div class="panic-installer-sub-chain__heading">
        <svc-icon color="primary" name="link" size="100px"></svc-icon>
        <svc-label class="panic-installer-sub-chain__title" position="center">
          Blockchain Setup
        </svc-label>
        <svc-label class="panic-installer-sub-chain__step" position="center">
          step 1/5
        </svc-label>
      </div>
      <div class="panic-installer-sub-chain__message-container">
        <p class="panic-installer-sub-chain__message">
          PANIC works with the concept of "base chains" and "sub-chains".
        </p>
        <svc-buttons-container>
          <svc-button color="secondary" iconname="information-circle" iconposition="icon-only"></svc-button>
        </svc-buttons-container>
      </div>
      <p class="panic-installer-sub-chain__message">
        <ul class="panic-installer-sub-chain__list">
          <li class="panic-installer-sub-chain__list-item">
            Supported base chains: Cosmos-SDK, Substrate, Chainlink or General - Used to represent any systems or repositories not associated with a blockchain. For example, the host machine that PANIC is running on.
          </li>
          <li>
            Sub-chains are any protocol built on top of the base chains. For example, Akash (Cosmos-SDK), Polkadot (Substrate).
          </li>
        </ul>
      </p>
      <form>
        <svc-select header="Base-chains" id="base_chain" name="base_chain" placeholder="Select a base chain..." withborder=""></svc-select>
        <svc-input id="subChainName" label="Name" lines="full" name="name" placeholder="Ex: Akash, Polkadot" required="" type="text"></svc-input>
        <svc-button id="submit" type="submit"></svc-button>
      </form>
    </svc-surface>
    <panic-installer-nav></panic-installer-nav>
  </svc-content-container>
  <panic-footer></panic-footer>
</panic-installer-sub-chain>
`;
